### ISSUE TITLE
Use ReflowMarkdown with 'Format Document' command

### ISSUE THREAD

**Issue by sbliven** [marvhen/ReflowMarkdown#7]
It would be nice to hook ReflowMarkdown to the 'Format Document' command so that all paragraphs get rewrapped. It appears this would require registering the extension as a markdown formatter.

**Comment by marvhen** [comment]
great idea! thanks!

**Comment by HeyItsGilbert** [comment]
I would definitely love to reflow my who document on save. Looks like it just requires registering with the right provider. https://code.visualstudio.com/blogs/2016/11/15/formatters-best-practices

Are you taking PR @marvhen  ?

**Comment by marvhen** [comment]
@HeyItsGilbert, I don't have a lot of free time to spend on this project right now, but if you have a working solution, I'll try to get it published.  I think the last time I tried to setup a dev/build environment I into issues trying to be compatible with older and newer versions of VSCode....I don't recall for sure.

**Comment by HeyItsGilbert** [comment]
I'll see what I can put together. Do you have a set of versions you want to ensure compatibility for? I'll see if there's something I can set on the build side to ensure we have backwards compatibility.

**Comment by HeyItsGilbert** [comment]
I see what you're saying about trying to upgrade the dev environment. Since TS is not in my usual repertoire I may bow out of trying to implement this. I may try to start from scratch and see where that ends up. If I get something working I'll let you know and start trying to make sure it's backwards compatible.

**Comment by HeyItsGilbert** [comment]
@marvhen I just got a new extension setup and started adding the code for formatting a file. And then I noticed you got 3.0.0 pushed lol. I'll submit a PR for just formatting the the file.

**Comment by marvhen** [comment]
Yeah, I decided to not really worry too much about backward compatibility.  Kinda taking a chance but VS Code is pretty stable with regular updates.  I have not gotten a lot of feature requests at all so I decided to try to accommodate those who are asking for them and those who don't want to risk it can easily downgrade.

**Comment by marvhen** [comment]
One thought about formatting the whole document: I think there will be some documents where it won't be good.  Users will just need to decide if it's really what they want because I know for me there are some paragraphs that I write that I specifically don't want to be reflowed.  Perhaps we can just comment on this in the readme...eventually we could consider a flag of some sort embedded in a comment or something similar to ignore certain lines when reflowing.

**Comment by HeyItsGilbert** [comment]
So in the PR is make sure to skip possible frontmatter and the config has `formatOnSave` as false by default.

I do think looking for something in line would be a good idea. Maybe we could have an include/exclude setting. Something similar to markdownlint?

**Comment by wilhelmer** [comment]
Thanks for your PR @HeyItsGilbert, I tried it and noticed that it needs a lot more modifications to make it work, and it changes the scope of the entire extension quite a bit.

So I created [Reflow Markdown Plus](https://marketplace.visualstudio.com/items?itemName=LWilhelmer.reflow-markdown-plus). It's very much tailored to my personal use case with Docusaurus, so might not be for everyone. But it's working great for me, I really like how the paragraphs reflow on every save while I'm working on a file.

@marvhen If you want to incorporate the features and drop the separate extension, happy to help.

**Comment by marvhen** [comment]
@wilhelmer sounds great! do you want to send a PR? I haven't looked at this repo in a while, but if you forked from it and got it working we should be good, no?

**Comment by wilhelmer** [comment]
Sorry for the late reply, I was on vacation 🌴 

I created a [PR](https://github.com/marvhen/ReflowMarkdown/pull/12) that includes all the features of the Plus version.
